### PR TITLE
SONAR-9802 New System Info page for cluster mode

### PR DESCRIPTION
**Pull request by gregaubert**

### CODE REVIEW

File: server/sonar-web/src/main/js/apps/system/components/info_items/ObjectItem.tsx [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+import * as React from 'react';
+import { map } from 'lodash';
+import ValueItem from './ValueItem';
+import { SysValueObject } from '../../types';
+
+interface Props {
+  value: SysValueObject;
+}
+
+const ObjectItem: React.StatelessComponent<Props> = ({ value }) => {
```
Comment (stas-vilchik) [inline server/sonar-web/src/main/js/apps/system/components/info_items/ObjectItem.tsx]: just use `export default function Foo`

File: server/sonar-web/src/main/js/apps/system/components/info_items/Section.tsx [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  return (
+    <div className="big-spacer-bottom">
+      <h3 className="spacer-bottom">{name}</h3>
+      <table className="data zebra" id={name}>
+        <tbody>
+          {map(items, (value, name) => {
+            return (
+              <tr key={name}>
+                <td className="thin">
+                  <div style={{ width: '25vw', overflow: 'hidden', textOverflow: 'ellipsis' }}>
```
Comment (stas-vilchik) [inline server/sonar-web/src/main/js/apps/system/components/info_items/Section.tsx]: please extract inline styles to a file

File: server/sonar-web/src/main/js/apps/system/components/info_items/BooleanItem.tsx [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+import * as React from 'react';
+
+interface Props {
+  value: boolean;
+}
+
+export default function BooleanItem({ value }: Props) {
```
Comment (stas-vilchik) [inline server/sonar-web/src/main/js/apps/system/components/info_items/BooleanItem.tsx]: rename directory `info_items` to `info-items`

File: server/sonar-web/src/main/js/components/icons-components/OpenCloseIcon.tsx [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+import * as React from 'react';
+
+interface Props {
+  className?: string;
+  open: boolean;
+  size?: number;
+}
+
+export default function OpenCloseIcon({ className, open, size = 10 }: Props) {
+  /* eslint-disable max-len */
```
Comment (stas-vilchik) [inline server/sonar-web/src/main/js/components/icons-components/OpenCloseIcon.tsx]: this is not needed

File: server/sonar-web/src/main/js/components/icons-components/OpenCloseIcon.tsx [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+  return (
+    <svg
+      className={className}
+      xmlns="http://www.w3.org/2000/svg"
+      viewBox="0 0 16 16"
+      width={size}
+      height={size}
+      style={{ fill: 'currentColor', paddingTop: '2px' }}>
```
Comment (stas-vilchik) [inline server/sonar-web/src/main/js/components/icons-components/OpenCloseIcon.tsx]: I suggest to not deal with paddings, but change the image itself, so it's 14x14 (or 16x16) pixels

File: server/sonar-web/src/main/js/api/system.ts [hunk 6]
```diff
@@ -24,7 +24,7 @@ export function setLogLevel(level: string): Promise<void> {
 }
 
 export function getSystemInfo(): Promise<any> {
```
Comment (stas-vilchik) [inline server/sonar-web/src/main/js/api/system.ts]: please add a response type, `SysInfo` if I'm correct

File: server/sonar-web/src/main/js/apps/system/components/ClusterSysInfos.tsx [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+import { sortBy } from 'lodash';
+import HealthCard from './info_items/HealthCard';
+import { translate } from '../../../helpers/l10n';
+import { SysInfo } from '../types';
+
+interface Props {
+  sysInfoData: SysInfo;
+}
+
+export default class ClusterSysInfos extends React.PureComponent<Props> {
```
Comment (stas-vilchik) [inline server/sonar-web/src/main/js/apps/system/components/ClusterSysInfos.tsx]: can be function

File: server/sonar-web/src/main/js/apps/system/components/info_items/HealthCard.tsx [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+interface Props {
+  biggerHealth?: boolean;
+  health: HealthType;
+  healthCauses: HealthCause[];
+  open: boolean;
+  name: string;
+}
+
+export default class HealthCard extends React.PureComponent<Props> {
```
Comment (stas-vilchik) [inline server/sonar-web/src/main/js/apps/system/components/info_items/HealthCard.tsx]: can be function

File: server/sonar-web/src/main/js/apps/system/components/App.tsx [hunk 9]
```diff
@@ -18,30 +18,52 @@
...
   mounted: boolean;
-  state: State = { loading: true };
+
+  static contextTypes = {
+    router: PropTypes.object
+  };
+
+  constructor(props: Props) {
+    super(props);
+    this.state = { loading: true, query: parseQuery(props.location.query) };
```
Comment (stas-vilchik) [inline server/sonar-web/src/main/js/apps/system/components/App.tsx]: I think it's not expensive to parse query each time it is used, so you can avoid putting it in `state`, and `componentWillReceiveProps` can be removed

File: server/sonar-web/src/main/js/components/common/RestartForm.tsx [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+  handleCancelClick = (event: React.SyntheticEvent<HTMLElement>) => {
+    event.preventDefault();
+    this.props.onClose();
+  };
+
+  handleFormSubmit = (event: React.SyntheticEvent<HTMLFormElement>) => {
+    event.preventDefault();
+    if (!this.state.restarting) {
+      this.setState({ restarting: true });
+      restartAndWait().then(() => document.location.reload(), throwGlobalError);
```
[the remark this answers is not in the document]
Comment (gregaubert) [reply server/sonar-web/src/main/js/components/common/RestartForm.tsx]: Ah yeah! I always forget!

**Comment by gregaubert** [comment]
@stas-vilchik could you review the last commit please ?

File: server/sonar-web/src/main/js/apps/system/components/info-items/HealthItem.tsx [hunk 11]
```diff
@@ -19,31 +19,24 @@
...
-          <span
-            key={idx}
-            className={classNames(
-              'spacer-right alert',
-              health === HealthType.RED ? 'alert-danger' : 'alert-warning'
-            )}>
-            {cause.message}
-          </span>
+      {hasHealthCauses &&
+        (healthCauses as HealthCause[]).map((cause, idx) => (
```
Comment (stas-vilchik) [inline server/sonar-web/src/main/js/apps/system/components/info-items/HealthItem.tsx]: instead of type-cast, use can use `healthCauses!.map(...)`

Comment (gregaubert) [reply server/sonar-web/src/main/js/apps/system/components/info-items/HealthItem.tsx]: :+1: I didn't know there was something like that! Thanks!

**Pull request merged**